Change Assimp to Silk.NET.Assimp implementation

Jklawreszuk opened the pull request:
# PR Details

This PR completes [johang88](https://github.com/johang88) attempt to migrate from native C++ Assimp code to Silk.NET bindings ([link to his PR](https://github.com/stride3d/stride/pull/1158)).

## Changes

- I checked previous code and can tell it's a 1-1 recreation of the native MeshConverter. However `ProcessMeshMaterial()` was incomplete so I wrote missing part. 
- Also, my PR includes small changes that simplify MeshConverter and changes proposed by other contributors (those I could understand)

Overall, I have not noticed any regression or difference in the import of objects, but I would be happy to hear any feedback :) 

Once, merged, I believe it will unlock the possibility of further development of the asset compiler for platforms other than Windows and work on glTF.

## Related Issue

#1394 
#360 

- [ ] My change requires a change to the documentation.
- [ ] I have added tests to cover my changes.
- [x] Breaking change (fix or feature that would cause existing functionality to change).

Ethereal77 commented on sources/tools/Stride.Importer.Assimp/MeshConverter.cs:
```diff
@@ -0,0 +1,1375 @@
+                for (var boneId = 0; boneId < nbBoneByVertex; ++boneId)
+                    totalWeight += curVertexWeights[boneId].Item2;
+
+                if (totalWeight == 0.0) // Assimp weights are positive, so in this case all weights are nulls
```
Floating point comparison. Is it safe to compare to 0 instead of an epsilon? I assume we expect it to be exactly zero when all weights are exactly 0 and no some combination of very low weights.

Ethereal77 commented on sources/tools/Stride.Importer.Assimp/MeshConverter.cs:
```diff
@@ -0,0 +1,1375 @@
+                {
+                    computeColorNode = new ComputeColor(baseColor);
+                }
+                //else if (hasBaseValue)
+                //{
+                //	computeColorNode = gcnew MaterialFloatComputeNode(baseValue);
+                //}
```
Why is this commented out?

NicusorN5 replied:
I think that's commented C++/CLI code, since C# doesn't have a `gcnew` keyword. `gcnew` is specific to C++/CLI and C++/CX.

Ethereal77 replied:
`gcnew` in C++/CLI is just the `new` keyword in C#, i.e. instantiating a new instance managed by the Garbage Collector.

NicusorN5 replied:
Yes, but this is a C# file, so there's no use for it. Maybe used as some sort of reference?

Ethereal77 replied:
Right now I'm not very versed in the material system. Maybe @tebjan knows better, but if there is a `MaterialFloatComputeNode` (or there were one) that maybe is a `IComputeValue` (does that type still exist?) it is/was useful somehow.
Or maybe it is merged into `ComputeColor` at this point and we should just create a regular `ComputeColor` with a single scalar value?
Anyway, that `else if (hasBaseValue)` indicates there was an use for it.

Ethereal77 replied:
Then must be a half-implemented feature, or a remnant from some old code. As I said, I'm not very knowledgeable about the material system, so don't know what would be the uses of that even if it sounds useful somehow.

Jklawreszuk replied:
I found original commit. Maybe It will be usefull for someone in the future: https://github.com/SiliconStudio/xenko//commit/d5ee29ea8efeceaa6eaf777ff903e4514f28c6f7

Ethereal77 commented on sources/tools/Stride.Importer.Assimp/MeshConverter.cs:
```diff
@@ -0,0 +1,1375 @@
+
+                finalMaterial.Attributes.Diffuse = new MaterialDiffuseMapFeature(computeColorNode);
+
+                // TODO TEMP: Set a default diffuse model
+                finalMaterial.Attributes.DiffuseModel = new MaterialDiffuseLambertModelFeature();
```
Why a default fixed Lambertian diffuse model?

Ethereal77 commented on sources/tools/Stride.Importer.Assimp/MeshConverter.cs:
```diff
@@ -0,0 +1,1375 @@
+                };
+                finalMaterial.Attributes.Specular = specularFeature;
+
+                // TODO TEMP: Set a default specular model
+                var specularModel = new MaterialSpecularMicrofacetModelFeature
+                {
+                    Fresnel = new MaterialSpecularMicrofacetFresnelSchlick(),
+                    Visibility = new MaterialSpecularMicrofacetVisibilityImplicit(),
+                    NormalDistribution = new MaterialSpecularMicrofacetNormalDistributionBlinnPhong()
+                };
+                finalMaterial.Attributes.SpecularModel = specularModel;
```
Why a default fixed Blinn-Phong specular model?

Ethereal77 commented on sources/tools/Stride.Importer.Assimp/MeshConverter.cs:
```diff
@@ -0,0 +1,1375 @@
+            }
+            else if (textureType == TextureType.TextureTypeEmissive)
+            {
+                // TODO: Add support
```
Would be nice to support all these texture map types, as they are pretty much standard today for PBR workflows.

tebjan replied:
Yes, fully agree, importing textures with the model would be great.

Jklawreszuk replied:
It looks like there is already an issue for this: https://github.com/stride3d/stride/issues/473

Ethereal77 reviewed (commented):
Good job. Hopefully we can also get rid of the C++ FBX importer in the near future.

I've commented some things to fix, and other things that should be good to note for future enhancements or refactorings.

NicusorN5 commented on sources/tools/Stride.Importer.Assimp/MeshConverter.cs:
```diff
@@ -0,0 +1,1375 @@
+                        if (float.IsNaN(tangentPointer->X) || float.IsNaN(tangentPointer->Y) || float.IsNaN(tangentPointer->Z) ||
+                            float.IsNaN(bitangentPointer->X) || float.IsNaN(bitangentPointer->Y) || float.IsNaN(bitangentPointer->Z))
+                        {
+                            //assert(mesh->HasNormals());
```
Assimp allows generation of normals, tangents and bitangents in the import settings by Using either the `aiProcess_GenSmoothNormals` or `aiProcess_GenNormals` importer flags.  From there, `aiProcess_CalcTangentSpace` can be used to generate tangents.

Not sure if normal and tangent space generation is wanted while importing a model.

Ethereal77 replied:
Then uncomment the assert, as it is assumed he must have normals at that point.
But also, what if you want a model / mesh without normals? Or just with normals but without tangent / bitangent? I'm specifically thinking about mesh colliders for example.

Ethereal77 commented:
I think if you add [WIP] to the title it is enough. But don't know.

Jklawreszuk commented on sources/tools/Stride.Importer.Assimp/Material/Materials.cs:
```diff
@@ -0,0 +1,154 @@
+        public static readonly Operation[] ConvertAssimpStackOperationCppToCs = new Operation[]
+        {
+            Operation.Add,			//aiStackOperation_Add
+		    Operation.Add3ds,		//aiStackOperation_Add3ds
```
Indeed weird. I guess github wrongly displays this file because when you open file everything is correct

Ethereal77 replied:
Maybe the indentation is not normalized between tabs and spaces? I tend to favor spaces because this very reason.

Jklawreszuk replied:
Yes this files had mixed tabs, Thanks!

Jklawreszuk commented on sources/tools/Stride.Importer.Assimp/Material/Materials.cs:
```diff
@@ -0,0 +1,154 @@
+        public static readonly StackType[] ConvertAssimpStackTypeCppToCs = new StackType[]
+        {
+            StackType.Color,			// aiStackType_ColorType
+		    StackType.Texture,		// aiStackType_TextureType
```
same as above

Jklawreszuk commented on sources/tools/Stride.Importer.Assimp/Material/Materials.cs:
```diff
@@ -0,0 +1,154 @@
+        public static readonly MappingMode[] ConvertAssimpMappingModeCppToCs = new MappingMode[]
+        {
+            MappingMode.Wrap,		// aiTextureMapMode_Wrap
+		    MappingMode.Clamp,		// aiTextureMapMode_Clamp
```
same as above

Jklawreszuk commented on sources/tools/Stride.Importer.Assimp/Utils.cs:
```diff
@@ -0,0 +1,66 @@
+            return (uint)n;
+        }
+
+        public static CompressedTimeSpan AiTimeToXkTimeSpan(double time, double aiTickPerSecond)
```
Good call!

VaclavElias commented:
Regarding the spaces, https://github.com/stride3d/stride/blob/master/.editorconfig is present, but not sure if it is in the right location?

Jklawreszuk commented on sources/tools/Stride.Importer.Common/TextureLayerGenerator.cs:
```diff
@@ -0,0 +1,61 @@
+            var uvSetName = "TEXCOORD";
+            if (textureUVSetIndex != 0)
+                uvSetName += textureUVSetIndex;
+            //albedoMaterial->Add(gcnew ShaderClassSource("TextureStream", uvSetName, "TEXTEST" + uvSetIndex));
```
Yea, It looks like in TextureLayerGenerator cpp class this line was commented out

Ethereal77 replied:
As it references things that no longer exist and also has the name "TEXTEST", we should remove it?

Jklawreszuk replied:
I am really against commenting out the code and then leaving it for a long time, so sure.

Ethereal77 commented:
To avoid this kind of problems with tabs and spaces I like to use [this extension](https://marketplace.visualstudio.com/items?itemName=VisualStudioPlatformTeam.FixMixedTabs) from Microsoft Labs. Highly recommended.

Ethereal77 commented:
Yeah, of course. My review is only suggestions to point to typos, indentation, comments, etc. Then comments about what we should aim for in a future PR.

Doprez commented:
> Question: should this be marked as a draft? There's lots of TO-DOs.

Although I havent cross referenced anything myself, there were a bunch of TO-DOs in the C++ files as well.

Jklawreszuk commented on sources/tools/Stride.Importer.Assimp/Utils.cs:
```diff
@@ -0,0 +1,66 @@
+            => new Color(v.X, v.Y, v.Z, v.W);
+
+        public static Core.Mathematics.Quaternion ToStrideQuaternion(this System.Numerics.Quaternion q)
+        // TODO: Not sure why I have to do this Assimp & System.Numerics.Quaternion seems to have the same memory layout (W, X, Y, Z)
+        // so just passing X, Y, Z, W without swizzling should work ...
+            => new Core.Mathematics.Quaternion(q.Y, q.Z, q.W, q.X); 
```
I have no idea too. After updating Silk package to latest version they introduced AssimpQuaternion structure. Also original implementation does not doing any swizzling...

Ethereal77 replied:
Maybe it is not needed anymore, or were a bug in the bindings and it is now corrected.

Ethereal77 commented:
Why have you closed the PR?

Jklawreszuk commented:
I am sorry, I force pushed the wrong branch on the remote by mistake and PR was automaticaly closed 😢
At least I have a lesson to avoid that.

Jklawreszuk commented:
Alright, so I refactored and moved the classes to separate files and applied all the review suggestions I could.
I think my changes are ready to be reviewed again.

Ethereal77 reviewed (commented):
Ok. Some more comments. I hope you don't hate me too much for some of these ones 😁.
The indentation one about tabs vs spaces is just a hassle. I favor spaces because render the same in every editor (unlike tabs, which render depending on editor/user config). But I understand it is a matter of preference.
As Stride has no coding style guidelines defined formally anywhere apart from imitating what you can find in other files or the rules of `.editorconfig`, it's up to you.
I don't want to be an indentation nazi 😀.

Jklawreszuk commented on sources/tools/Stride.Importer.Assimp/Material/Flags.cs:
```diff
@@ -0,0 +1,19 @@
+﻿// Copyright (c) .NET Foundation and Contributors (https://dotnetfoundation.org/ & https://stride3d.net) and Silicon Studio Corp. (https://www.siliconstudio.co.jp)
```
Hmm...I am not sure about that. Because I originaly separated classes/enums from this file:
https://github.com/stride3d/stride/blob/master/sources/tools/Stride.Assimp/MaterialStack.cs

Ethereal77 replied:
If it comes from another file with a license header, forget about my comment then. For other files with no header or for new files, however, use the "simple" one.

Jklawreszuk commented:
> Ok. Some more comments. I hope you don't hate me too much for some of these ones 😁.

@Ethereal77 Not at all. That's the purpose of code review, to make sure everything is alright 😅 
Thanks again for all suggestions. I think that's the good point to have identation consistency, so I changed for every file to space instead of tabs. Would be also nice to take care of .editorconfig file in future. Other than that I think reviewed again (I hope didn't miss anything 😂)

VaclavElias commented:
Nice attitude 🙂

Ethereal77 commented on sources/tools/Stride.Importer.Assimp/Material/Flags.cs:
```diff
@@ -0,0 +1,18 @@
+namespace Stride.Importer.Assimp.Material
+{
+    /// <summary>
+    /// Enumeration of the new Assimp's flags.
```
A better explanation of this types would be nice.
It's ok however if you are not very familiar with graphics and/or meshes/Assimp. That is something that can be later expanded in a PR to improve documentation I think.

Ethereal77 reviewed (commented):
LGTM.
This is a first good step towards a more developed content pipeline for Stride.

Jklawreszuk commented:
If there are no objections, I think changes are ready to merge 😊

Eideren commented:
Thanks a bunch for taking the time to look at this @Ethereal77, and @NicusorN5 for providing additional insight !

I'm guessing this one will be taken care of in another issue/pr @Jklawreszuk ?
https://github.com/stride3d/stride/pull/1719/files#r1285751424

Eideren commented:
And of course, thank you for going through this stuff @Jklawreszuk

Jklawreszuk commented:
> Thanks a bunch for taking the time to look at this @Ethereal77, and @NicusorN5 for providing additional insight !
> 

Also, we should thank @johang88 for his big effort 🫡🫡 . I'm impressed. 

> I'm guessing this one will be taken care of in another issue/pr @Jklawreszuk ?

Definitely, we need now to discuss what to do next with FBX importer, because having cpp compiler as requirement during building is pain and I think potential PR would cover additional changes to assimp importer.
